server/src/test/java/org/elasticsearch/cluster/metadata/DiffableStringMapTests.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.cluster.Diff;
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.instanceOf;
+
+public class DiffableStringMapTests extends ESTestCase {
+
+    public void testDiffableStringMapDiff() {
+        Map<String, String> m = new HashMap<>();
+        m.put("foo", "bar");
+        m.put("baz", "eggplant");
+        m.put("potato", "canon");
+        DiffableStringMap dsm = new DiffableStringMap(m);
+
+        Map<String, String> m2 = new HashMap<>();
+        m2.put("foo", "not-bar");
+        m2.put("newkey", "yay");
+        m2.put("baz", "eggplant");
+        DiffableStringMap dsm2 = new DiffableStringMap(m2);
+        logger.info("--> 1: {}", dsm);
+        logger.info("--> 2: {}", dsm2);
+
+        Diff<DiffableStringMap> diff = dsm2.diff(dsm);
+        assertThat(diff, instanceOf(DiffableStringMap.DiffableStringMapDiff.class));
+        DiffableStringMap.DiffableStringMapDiff dsmd = (DiffableStringMap.DiffableStringMapDiff) diff;
+
+        assertThat(dsmd.getDeletes(), containsInAnyOrder("potato"));
+        assertThat(dsmd.getDiffs().size(), equalTo(0));
+        Map<String, String> upserts = new HashMap<>();
+        upserts.put("foo", "not-bar");
+        upserts.put("newkey", "yay");
+        assertThat(dsmd.getUpserts(), equalTo(upserts));
+
+        DiffableStringMap dsm3 = diff.apply(dsm);
+        logger.info("--> 3: {}", dsm3);
+        assertThat(dsm3.get("foo"), equalTo("not-bar"));
+        assertThat(dsm3.get("newkey"), equalTo("yay"));
+        assertThat(dsm3.get("baz"), equalTo("eggplant"));
+        assertThat(dsm3.get("potato"), equalTo(null));
+    }
```

Comment:
Can we also add a test here with a few more maps and entries (maybe randomly generated) than in the existing test?

Reply:
and maybe also a test that verifies that the serialization works?

Reply:
I also wonder if we need a random test that adds randomly to one map and then passes it on as a diff to another and makes sure after we did that N times the maps are the same?

Reply:
> and maybe also a test that verifies that the serialization works?

The serialization was added as part of the `IndexMetaDataTests`, but I'll add one here for targeted serialization

Reply:
> I also wonder if we need a random test that adds randomly to one map and then passes it on as a diff to another and makes sure after we did that N times the maps are the same?

I'll add a test for that as well